test/singa/test_image_transformer.cc

```diff
@@ -0,0 +1,261 @@
+        // test for HWC
+        size_t in_idx = ((h + 2) * width + 3 + w) * channel + c,
+              out_idx = (h * 3 + w) * channel + c;
+        EXPECT_EQ(x[in_idx], y[out_idx]);
+      }
+#endif
+  delete[] x;
+}
+
+TEST(ImageTransformer, Apply2D) {
+  size_t n = 60;
+  float* x = new float[n];
+  size_t height = 6, width = 10;
+  singa::Tensor in(singa::Shape{height, width});
+  srand(time(NULL));
+  for (size_t i = 0; i < n; i++) x[i] = (float)(rand() % 256);
+  in.CopyDataFromHostPtr<float>(x, n);
+  size_t resize_height = 4, resize_width = 6;
+
+  singa::ImageTransformer img_transformer;
+  singa::TransformerConf conf;
+  conf.set_resize_height(resize_height);
+  conf.set_resize_width(resize_width);
+  conf.set_horizontal_mirror(false);
+  conf.set_image_dim_order("HWC");
+  conf.add_crop_shape(2u);
+  conf.add_crop_shape(3u);
+  img_transformer.Setup(conf);
+
+  singa::Tensor out = img_transformer.Apply(singa::kEval, in);
+  EXPECT_EQ(2u, out.shape(0));
+  EXPECT_EQ(3u, out.shape(1));
+  const float* y = out.data<float>();
+#ifdef USE_OPENCV
+  cv::Mat mat(height, width, CV_32FC1, cv::Scalar(0, 0, 0));
+  for (size_t i = 0; i < height; i++)
+    for (size_t j = 0; j < width; j++)
+      mat.at<cv::Vec<float, 1>>(i, j)[0] = x[i * width + j];
+  cv::Size size(resize_width, resize_height);
+  cv::Mat resized;
+  cv::resize(mat, resized, size);
+  EXPECT_EQ(resize_height, resized.size().height);
+  EXPECT_EQ(resize_width, resized.size().width);
+  size_t new_size = resize_height * resize_width;
+  float* xt = new float[new_size];
+  for (size_t i = 0; i < resize_height; i++)
+    for (size_t j = 0; j < resize_width; j++)
+        xt[i * resize_width + j] = resized.at<cv::Vec<float, 1>>(i, j)[0];
+
+  for (size_t h = 0; h < 2; h++)
+    for (size_t w = 0; w < 3; w++){
+      size_t in_idx = (h + 1) * resize_width + 1 + w,
+            out_idx = h * 3 + w;
+      EXPECT_EQ(xt[in_idx], y[out_idx]);
+    }
+  delete[] xt;
+#else
+  for (size_t h = 0; h < 2; h++)
+    for (size_t w = 0; w < 3; w++){
+      size_t in_idx = (h + 2) * width + 3 + w,
+            out_idx = h * 3 + w;
+      EXPECT_EQ(x[in_idx], y[out_idx]);
+    }
+#endif
+  delete[] x;
+}
+
+#ifdef USE_OPENCV
+TEST(ImageTransformer, Resize) {
+  size_t n = 180;
+  float* x = new float[n];
+  size_t channel = 3, height = 6, width = 10;
+  singa::Tensor in(singa::Shape{height, width, channel});
+  srand(time(NULL));
+  for (size_t i = 0; i < n; i++) x[i] = (float)(rand() % 256);
+  in.CopyDataFromHostPtr<float>(x, n);
+  size_t resize_height = 4, resize_width = 5;
+  singa::Tensor out = singa::resize(in, resize_height, resize_width, "HWC");
+  const float* y = out.data<float>();
+
+  cv::Mat mat(height, width, CV_32FC3, cv::Scalar(0, 0, 0));
+  for (size_t i = 0; i < height; i++)
+    for (size_t j = 0; j < width; j++)
+      for (size_t k = 0; k < channel; k++)
+        mat.at<cv::Vec3f>(i, j)[k] = x[i * width * channel + j * channel + k];
```

Comment:
TODO better check the datatype of the tensors, uint8 or float.
Tensor::AsType could be helpful for data type change.